engine/src/main/java/org/hibernate/search/engine/impl/WorkPlan.java

```diff
@@ -111,12 +111,36 @@ public int size() {
 	 * @param typeIdentifier The entity class for which to retrieve the work
 	 *
 	 * @return returns (and creates if needed) the {@code PerClassWork} from the {@link #byClass} map.
+	 * @throws SearchException if the specified type is not @Indexed nor hosts any @ContainedIn
 	 */
 	private PerClassWork getClassWork(String tenantId, IndexedTypeIdentifier typeIdentifier) {
+		PerClassWork classWork = getClassWorkIfConfigured( tenantId, typeIdentifier );
+		if ( classWork == null ) {
+			throw new SearchException(
+					"Unable to perform work. Entity Class is not @Indexed nor hosts @ContainedIn: " + typeIdentifier
+			);
+		}
+		return classWork;
+	}
+
+	/**
+	 * @param tenantId the tenant identifier
+	 * @param typeIdentifier The entity class for which to retrieve the work
+	 *
+	 * @return the {@code PerClassWork} from the {@link #byClass} map, creating it if necessary and if possible
+	 * (if the type is configured). Otherwise, return null.
+	 */
+	private PerClassWork getClassWorkIfConfigured(String tenantId, IndexedTypeIdentifier typeIdentifier) {
```

Comment:
I suppose there's a good reason you didn't use an `Optional` here? It would make a lot of sense.

Reply:
Mainly because we don't use such features a lot in Search 5. But now that I think of it, it also makes the code more painful to backport to 5.6, which uses Java 7.

Reply:
Right, makes sense.